Mathlib/MeasureTheory/MeasurableSpace.lean

```diff
@@ -1664,42 +1665,36 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
     CountablyGenerated (α × β) :=
   .sup (.comap Prod.fst) (.comap Prod.snd)
 
+instance [MeasurableSpace α] {s : Set α} [h : CountablyGenerated s] [MeasurableSingletonClass s] :
+    HasCountableSeparatingOn α MeasurableSet s := by
+  suffices HasCountableSeparatingOn s MeasurableSet univ from this.of_subtype fun _ ↦ id
+  rcases h.1 with ⟨b, hbc, hb⟩
+  refine ⟨⟨b, hbc, fun t ht ↦ hb.symm ▸ .basic t ht, fun x _ y _ h ↦ ?_⟩⟩
+  rw [← forall_generateFrom_mem_iff_mem_iff, ← hb] at h
+  simpa using h {y}
+
 variable (α)
 
 open Classical
 
 /-- If a measurable space is countably generated, it admits a measurable injection
 into the Cantor space `ℕ → Bool` (equipped with the product sigma algebra). -/
 theorem measurable_injection_nat_bool_of_countablyGenerated [MeasurableSpace α]
-    [h : CountablyGenerated α] [MeasurableSingletonClass α] :
+    [h : HasCountableSeparatingOn α MeasurableSet univ] :
     ∃ f : α → ℕ → Bool, Measurable f ∧ Function.Injective f := by
-  obtain ⟨b, bct, hb⟩ := h.isCountablyGenerated
-  obtain ⟨e, he⟩ := Set.Countable.exists_eq_range (bct.insert ∅) (insert_nonempty _ _)
-  rw [← generateFrom_insert_empty, he] at hb
-  refine' ⟨fun x n => x ∈ e n, _, _⟩
+  rcases h.1 with ⟨S, hSc, hSm, hS⟩
+  wlog hne : S.Nonempty generalizing S
+  · exact this (insert ∅ S) (hSc.insert ∅) (forall_insert_of_forall hSm .empty)
+      (fun x hx y hy h ↦ hS x hx y hy (forall_of_forall_insert h)) (insert_nonempty _ _)
```

Comment:
could you have this as a separate lemma in the `HasCountableSeparatingOn`: given a set `s` which satisfies `p s`, then there is a nonempty countable set `S` separating the points and satisfying `p`. And another lemma that gives this as the range of a function from nat?